File: 1/t2.py
def calc_grade(datas,grades,index):
    ld = len(datas)
    cost = 0
    if index == ld-1:
        pass
    else:
        for i in range(index+1,ld):
            cost += datas[i][0] - datas[index][0]
        for i in range(index+1,ld):
            mycost = grades[i]
            for j in range(index+1,i):
                mycost += datas[j][0] - datas[index][0]
            cost = min(mycost,cost)
    return cost + datas[index][1]
n = int(input())
ds = input().split()
ts = input().split()
datas = [(int(ds[i]),int(ts[i])) for i in range(n)]
datas.sort(key=lambda x:x[0])
grades = [0]*n
for i in range(len(datas)-1,-1,-1):
    grades[i] = calc_grade(datas,grades,i)
print(grades[0])

File: 1/t2.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;
#define MIN(a,b) (a<b?a:b)

struct table{
    int distance,cost,grade;
    int flood;
    bool operator < (const table& str) const
    {
        return (distance < str.distance);
    }
};
void calc_grade(vector<table> &tables,int index){
    int n;
    int flood,cost,distance;
    int grade;
    n = tables.size();
    if (index != n-1){
        flood = 1+tables[index+1].flood;
        cost = tables[index+1].cost;
        distance =  tables[index+1].distance-tables[index].distance;
        grade = tables[index+1].grade;
    }else{
        flood = 1;
        cost = 0;
        distance = 0;
        grade = 0;
    }
    tables[index].grade = grade + (flood - 1)*distance;
    if(index == 0||(tables[index].cost<=flood*(tables[index].distance - tables[index-1].distance))){
        tables[index].flood = 0;
        tables[index].grade += tables[index].cost;
    }else{
        tables[index].flood = flood;
    }
    //cout<< tables[index].grade<<"____"<<tables[index].flood<<endl;
}
int main(){
    int n;
    cin >> n;
    vector<table> tables(n);
    for(int i=0;i<n;i++){
        cin >> tables[i].distance;
    }
    for(int i=0;i<n;i++){
        cin >> tables[i].cost;
    }
    sort(tables.begin(),tables.end());
    for(int i=n-1;i>=0;i--){
        calc_grade(tables,i);
    }
    cout<<tables[0].grade<<endl;
}
